src/OrbitCaptureClient/GpuQueueSubmissionProcessor.cpp

```diff
@@ -103,23 +104,23 @@ const GpuQueueSubmission* GpuQueueSubmissionProcessor::FindMatchingGpuQueueSubmi
   auto lower_bound_gpu_submission_it =
       post_submission_time_to_gpu_submission.lower_bound(submit_time);
   if (lower_bound_gpu_submission_it == post_submission_time_to_gpu_submission.end()) {
-    return nullptr;
+    return std::nullopt;
   }
-  const GpuQueueSubmission* matching_gpu_submission = &lower_bound_gpu_submission_it->second;
+  const GpuQueueSubmission& matching_gpu_submission = lower_bound_gpu_submission_it->second;
 
-  if (matching_gpu_submission->meta_info().pre_submission_cpu_timestamp() > submit_time) {
-    return nullptr;
+  if (matching_gpu_submission.meta_info().pre_submission_cpu_timestamp() > submit_time) {
+    return std::nullopt;
   }
 
   return matching_gpu_submission;
 }
 
-const GpuJob* GpuQueueSubmissionProcessor::FindMatchingGpuJob(
+std::optional<GpuJob> GpuQueueSubmissionProcessor::FindMatchingGpuJob(
```

Comment:
In a way I'm still a bit confused. Was the map modified after this function was called? I'm saying that because now a copy is made, which is fine performance-wise, yet feels conceptually wrong.

Reply:
To be honest, I don't see where the map may get modified after I retrieved the element. But still this seemed to be the issue. Maybe some weird thing it the MS world...

Reply:
should probably take another look at this then?

Reply:
Okay, I found out what was the original issue and am actually a bit shocked how this never blown up on Linux and only occasionally on Windows.

When looking at this method: We extract a lot of information from the "completed marker", which is a member of the "gpu_queue_submission". Originally that was a pointer to an element inside the map.

Now, if that "completed" maker has information about a begin marker (the true case in line 312), we do some further processing. In particular we set the "matching_begin_submission" to the "gpu_queue_submission", if the meta data is equal. After we are done with the processing of the "matching_begin_submission" we call DecrementUnprocessedBeginMarkers, to eventually clean up state that we might not need anymore, this will (in some cases) remove the "matching_begin_submission" from the map, BUT, in this case it is equal to the "gpu_queue_submission", and this is used right after the method call, so right after it was erased....

I'll move reads of the "gpu_queue_submission" upfront, so that the erase is the last thing that would be done. WDYT?
Especially about std::option vs. pointer inside the std::map.

Reply:
I changed it, by doing the "Decrement.."  and thus the possible erase operation as last step.